File: JavaScript/countPatternsFrom.js
// Feel free to write and use any additional functions, variables, objects, etc. as you wish

const fullGrid = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I'];

const pairOfBlockingLetters = (letter) => {
  switch (letter) {
    case 'A':
      return [['B', 'C'], ['E', 'I'], ['D', 'G']];
    case 'B':
      return [['E', 'H']];
    case 'C':
      return [['B', 'A'], ['E', 'G'], ['F', 'I']];
    case 'D':
      return [['E', 'F']];
    case 'E':
      return [];
    case 'F':
      return [['E', 'D']];
    case 'G':
      return [['D', 'A'], ['E', 'C'], ['H', 'I']];
    case 'H':
      return [['E', 'B']];
    case 'I':
      return [['H', 'G'], ['E', 'A'], ['F', 'C']];
    default:
      return null;
  }
};

/* String, Int ->  Int */
const countPatternsFrom = (firstDot, length, grid = fullGrid) => {
  if (length < 1 || length > 9) return 0;
  if (length === 1) return 1;

  switch (firstDot) {
    case 'A': {
      const restOfGrid = grid.filter(x => x === 'A');
      let options = 0;
      pairOfBlockingLetters('A').forEach((pair) => {
        const [firstLetter, blockedLetter] = pair;
        if (restOfGrid.includes(firstLetter)) {
          options += countPatternsFrom(firstLetter, length - 1, restOfGrid);
        } else if (restOfGrid.includes(blockedLetter)) {
          options += countPatternsFrom(blockedLetter, length - 1, restOfGrid);
        }
      });
      ['F', 'G'].forEach((letter) => {
        if (restOfGrid.includes(letter)) {
          options += countPatternsFrom(letter, length - 1, restOfGrid);
        }
      });
      return options;
    }
    case 'B':
      break;
    case 'C':
      break;
    case 'D':
      break;
    case 'E':
      break;
    case 'F':
      break;
    case 'G':
      break;
    case 'H':
      break;
    case 'I':
      break;
    default:
      return null;
  }
  return null;
};

console.log(countPatternsFrom('A', 0)); // , 0);
console.log(countPatternsFrom('A', 10)); // , 0);
console.log(countPatternsFrom('B', 1)); // , 1);
console.log(countPatternsFrom('C', 2)); // , 5);

console.log(countPatternsFrom('D', 3)); // , 37);
console.log(countPatternsFrom('E', 4)); // , 256);
console.log(countPatternsFrom('E', 8)); // , 23280);

File: Ruby/count_patterns_from.rb
$full_grid =  ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I']

def pairs_of_blocking_letters(letter)
  case letter
  when 'A'
    return [['B', 'C'], ['E', 'I'], ['D', 'G']]
  when 'B'
    return [['E', 'H']]
  when 'C'
    return [['B', 'A'], ['E', 'G'], ['F', 'I']]
  when 'D'
    return [['E', 'F']]
  when 'E'
    return []
  when 'F'
    return [['E', 'D']]
  when 'G'
    return [['D', 'A'], ['E', 'C'], ['H', 'I']]
  when 'H'
    return [['E', 'B']]
  when 'I'
    return [['H', 'G'], ['E', 'A'], ['F', 'C']]
  else
    return letter
  end
end

def other_letters(letter)
  this_and_blockers = pairs_of_blocking_letters(letter).flatten + [letter]
  $full_grid.reject {|x| this_and_blockers.include?(x)}
end

def count_patterns_from(first_dot, length, grid = $full_grid)
  return 0 if !length.between?(1, 9)
  return 1 if length == 1

  rest_of_grid = grid.reject {|x| x == first_dot}
  options = 0

  pairs_of_blocking_letters(first_dot).each do |pair|
    first_letter, blocked_letter = pair
    if rest_of_grid.include?(first_letter)
      options += count_patterns_from(first_letter, length - 1, rest_of_grid)
    elsif rest_of_grid.include?(blocked_letter)
      options += count_patterns_from(blocked_letter, length - 1, rest_of_grid)
    end
  end

  other_letters(first_dot).each do |letter|
    if rest_of_grid.include?(letter)
      options += count_patterns_from(letter, length - 1, rest_of_grid)
    end
  end

  options
end

puts count_patterns_from('C', 1)
puts count_patterns_from('C', 2)
puts count_patterns_from('C', 3)
puts count_patterns_from('C', 4)